src/fsharp/FSharp.Core/printf.fs

```diff
@@ -1182,401 +935,288 @@ module internal PrintfImpl =
+                        StepStar1 (prefix, convFunc)
             else
-                buildPlainFinal(plainArgs, plainTypes)
-
-        let rec parseFromFormatSpecifier (prefix: string) (s: string) (funcTy: Type) i: int = 
+                // For interpolated string format processing, the static types of the '%A' arguments 
+                // are provided via CaptureTypes and are only known on second phase.
+                match argTys with
+                | null when spec.TypeChar = 'A' ->
+                    let convFunc arg argTy = 
+                        let mi = mi_GenericToString.MakeGenericMethod [| argTy |]
+                        let f = mi.Invoke(null, [| box spec |]) :?> ValueConverter
+                        let f2 = f.FuncObj :?> (obj -> string)
+                        f2 arg
+
+                    StepWithTypedArg (prefix, convFunc)
+
+                | _ -> 
+                    // For curried interpolated string format processing, the static types of the '%A' arguments 
+                    // are provided via the argument typed extracted from the curried function. They are known on first phase.
+                    let argTy = match argTys with null -> typeof<obj> | _ -> argTys.[0]
+                    let conv = getValueConverter argTy spec
+                    let convFunc = conv.FuncObj :?> (obj -> string)
+                    StepWithArg (prefix, convFunc)
             
-            if i >= s.Length then 0
-            else
+        let parseSpec i = 
+            let flags, i = FormatString.parseFlags fmt (i + 1)
+            let width, i = FormatString.parseWidth fmt i
+            let precision, i = FormatString.parsePrecision fmt i
+            let typeChar, i = FormatString.parseTypeChar fmt i
+            let interpHoleDotnetFormat, i = FormatString.parseInterpolatedHoleDotNetFormat typeChar fmt i
+
+            // Skip %P insertion points added after %d{...} etc. in interpolated strings
+            let i = FormatString.skipInterpolationHole fmt i
+
+            let spec = 
+                { TypeChar = typeChar
+                  Precision = precision
+                  Flags = flags
+                  Width = width
+                  InteropHoleDotNetFormat = interpHoleDotnetFormat }
+            i, spec
             
-            System.Diagnostics.Debug.Assert(s.[i] = '%', "s.[i] = '%'")
-            count <- count + 1
-
-            let flags, i = FormatString.parseFlags s (i + 1)
-            let width, i = FormatString.parseWidth s i
-            let precision, i = FormatString.parsePrecision s i
-            let typeChar, i = FormatString.parseTypeChar s i
-            let spec = { TypeChar = typeChar; Precision = precision; Flags = flags; Width = width}
+        // A simplified parser. For the case where the string is being used with interpolands captured in the Format object. 
+        let rec parseStepsAux steps (prefix: string) i = 
```

Comment:
Can this be changed to generate an array so we don't have to reverse it later?

Reply:
Done, though it still is accumulated as a list as we don't know the count.  We just reverse it into an array directly now